server/tests-py/test_upgrade/test_basics.py

```diff
@@ -0,0 +1,76 @@
+import ruamel.yaml as yaml
+import pytest
+from validate import check_query_f
+
+
+@pytest.mark.parametrize("transport", ['http', 'websocket'])
+class TestGraphQLQueryBasic():
```

Comment:
Could we avoid duplicating the definitions of these tests? It should be possible to specify particular tests to run given a single invocation of `pytest`, which is what we do in `test_server.sh` already.

Reply:
Another, although smaller, reason was that in our test suite, we have different `setup` and `teardown` for `TestGraphQLQueryBasic` and `TestGraphQLInsertBasic` and so on. Again, since we are avoiding setup and teardown all-together we would need to have a common migration which works for all these test classes.

Reply:
I see, that makes sense. (This is another good reason to move these tests to Haskell—it would likely be much easier to add the necessary abstraction in Haskell than in Python!)

One thing you could probably do is add a common superclass, like this:

```python
class GraphQLQueryBasicMixin():
  def test_select_query_author(self, hge_ctx, transport):
    ...
  ...

class TestGraphQLQueryBasic(DefaultTestSelectQueries, GraphQLQueryBasicMixin): pass
```

Then you could reuse the common superclass in the migration test. But to be honest, I don’t really like this, as it still requires the migrations to be manually kept in sync with the setup/teardown actions.

A better solution would be to abstract over things further so that you could reuse the definition of the setup action as the migration itself. You could do this, too—the tests are just code, like everything else!—but it would require a little more plumbing. I don’t think it would take too much effort to get sharing between the two paths, though.